tools/api-markdown-documenter/src/api-item-transforms/configuration/Hierarchy.ts

```diff
@@ -0,0 +1,407 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { type ApiItem, ApiItemKind, type ApiPackage } from "@microsoft/api-extractor-model";
+
+import {
+	getApiItemKind,
+	getUnscopedPackageName,
+	type ValidApiItemKind,
+} from "../../utilities/index.js";
+import { createQualifiedDocumentNameForApiItem } from "../ApiItemTransformUtilities.js";
+
+/**
+ * Kind of documentation suite hierarchy.
+ *
+ * @public
+ */
+export enum HierarchyKind {
+	/**
+	 * The API item gets a section under the document representing an ancestor of the API item.
+	 */
+	Section = "Section",
+
+	/**
+	 * The API item gets its own document, in the folder for an ancestor of the API item.
+	 */
+	Document = "Document",
+
+	/**
+	 * The API item gets its own document, and generates folder hierarchy for all descendent API items.
+	 */
+	Folder = "Folder",
+}
+
+/**
+ * {@link DocumentationHierarchyConfiguration} base interface.
+ *
+ * @remarks
+ * Not intended for external use.
+ * Only exists to share common properties between hierarchy configuration types.
+ *
+ * @sealed
+ * @public
+ */
+export interface DocumentationHierarchyConfigurationBase<THierarchyKind extends HierarchyKind> {
```

Comment:
Do we envision an extension of `HierarchyKind`? Something triggers my spidey-sense here, feels like too much boilerplate... would it be enough to make `readonly kind: "Section" | "Document" | "Folder"` and document all values on the property itself? Or point to the types for each kind.

Reply:
The main goal with this is to allow each of the implementations to implement exactly 1 kind such that type-narrowing works when switching on that property. But we don't _strictly_ need a base configuration for that. Each type could just have a `kind` property typed with the specific flag. That probably is a bit cleaner. I'll update.